extensions/amp-user-notification/0.1/amp-user-notification.js

```diff
@@ -0,0 +1,335 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Layout} from '../../../src/layout';
+import {assertHttpsUrl} from '../../../src/url';
+import {assert} from '../../../src/asserts';
+import {cidFor} from '../../../src/cid';
+import {getService} from '../../../src/service';
+import {log} from '../../../src/log';
+import {xhrFor} from '../../../src/xhr';
+
+
+const TAG_ = 'AmpUserNotification';
+
+/**
+ * @typedef {{
+ *   elementId: string,
+ *   ampUserId: string
+ * }}
+ */
+let GetRequestMetadata;
+
+/**
+ * @typedef {{
+ *   showNotification: boolean
+ * }}
+ */
+let GetResponseMetadata;
+
+/**
+ * @typedef {{
+ *   promise: (!Promise),
+ *   resolve: (function(*))
+ * }}
+ */
+let UserNotificationDefer;
+
+/**
+ * Component class that handles a simple notification.
+ */
+export class AmpUserNotification extends AMP.BaseElement {
+
+  /** @override */
+  isLayoutSupported(layout) {
+    return layout != Layout.FILL;
+  }
+
+  /** @override */
+  buildCallback() {
+    /** @private @const {!Window} */
+    this.win_ = this.getWin();
+
+    /** @private @const {!UserNotification} */
+    this.userNotification_ = getUserNotification_(this.win_);
+
+    /** @private {!Promise} */
+    this.activatedPromise_ = null;
+
+    assert(this.element.hasAttribute('id'), 'amp-user-notification should ' +
+        'have an id.');
+
+    /** @private {string} */
+    this.elementId_ = this.element.getAttribute('id');
+
+    /** @private {?string} */
+    this.ampUserId_ = null;
+
+    assert(this.element.hasAttribute('data-show-if-href'),
+        'amp-user-notification should have `data-show-if-href` attribute.');
+
+    /** @private @const {string} */
+    this.showIfHref_ = assertHttpsUrl(
```

Comment:
i think i did this because assertHttpsUrl has an exception for localhost and i wanted a test for it. i'll remove.